design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+
+The tool creates two metadata objects, a stream and table. automatically during the first migration; or when the user runs the tool with a parameter to initialize it.
+
+i.e.
+```
+$ ksql-migrations initialize
+Schema metadata initialized successfully
+```
+
+Due to some query limitations (lack of ORDER BY clause and pull queries working only on materialized views) in ksqlDB, the metadata will be stored in two places;  
+One stream (MIGRATION_EVENTS) and one table (SCHEMA_VERSION).
+
+The stream and table require topics unique for the cluster. In this case, these topics names will use the same convention as the processing log. It's not going
+to be an internal topic because it will be a user topic. The topic name is: `{clusterID}ksql_{StreamOrTableName}`.
+
+The user can specify a different name for the SCHEMA_VERSION table. This can be done through the tool configuration file. See `Configurations` for more details.
+
+The `MIGRATION_EVENTS` stream will keep track of every migration change (including undo changes). This will contain the history of changes the user has done. Also, this stream
+will contain a key to the current version of the schema (specified as `CURRENT` version). Every time a new migration or undo happens, the tool will insert a new event with the
+`CURRENT` key pointing to the current version.
+
+This is the `CREATE` statement for the `MIGRATION_EVENTS` stream:
+```sql
+CREATE STREAM migration_events (
+  version_key  STRING KEY,
+  version      STRING,
+  name         STRING,
+  state        STRING,  
+  checksum     STRING,
+  started_on   STRING,
+  completed_on STRING,
+  previous     STRING
+) WITH (  
+  KAFKA_TOPIC='default_ksql_migration_events',
+  VALUE_FORMAT='JSON',
+  PARTITIONS=1,
+  REPLICAS=1
+);
+```
+
+The `version_key` column has the version of the migration applied or undone. Special values of the `version_key`, `CURRENT` and `LATEST` will be reserved for internal purposes.
+The `version` column has the version of the migration applied.
+The `name` column has the name of the migration.
+The `state` column has the state of the migration process. It can be any of `Pending`, `Running`, `Migrated`, `Error`, `Undone`.
+The `checksum` column has the MD5 checksum of the migration file. It is used to validate the schema migrations with the local files.
+The `started_on` column has the date and time when the migration started.
+The `completed_on` column has the date and time when the migration finished.
+The `previous` column has the previous version applied.
+
+The `SCHEMA_VERSION` table will also keep track of every migration change (including future work for undo changes), but with the difference that being a table the tool will see quickly if a schema
+version has been migrated or undone. It will also give us a quick view of the `CURRENT` state of the schema. The major advantage is that the tool will use pull queries in this materialized
```

Comment:
I'm not sure about this.

I understand that inserting a row in the stream may eventually be available in the MV for a pull query. So my idea was this:
- Insert a row in the stream for a Pending migration (i.e. v1)
- Run the pull query on the MV and wait until CURRENT is set to v1 and Pending (fail otherwise)
- Execute the v1 migration script
- Insert a row in the stream to change Pending -> Migrated
- Run the pull query on the MV and wait until CURRENT is set to v1 and Migrated (fail otherwise)
- Continue with the next migration script

But you mentioned queries at different times and different replicas. I think even a complete state store clean up would cause this consistency issues because when the ksql node is restarted, it will take time to sync the state store again. Do you have some ideas how to do it?

Reply:
I don't really have a good idea about how to work around this until we have vector clock support (cc @vvcephei) so that we can ensure we're at least up to date as of a certain point